instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/TracingCallFactory.java

```diff
@@ -0,0 +1,121 @@
+package io.opentelemetry.instrumentation.okhttp.v3_0;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.caching.Cache;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okio.Timeout;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+class TracingCallFactory implements Call.Factory {
+  private static final Cache<Request, Context> contextsByRequest = Cache.newBuilder().setWeakKeys().build();
+  private final OkHttpClient okHttpClient;
+
+  TracingCallFactory(OkHttpClient okHttpClient) {this.okHttpClient = okHttpClient;}
+
+  @Nullable
+  static Context getCallingContextForRequest(Request request) {
+    return contextsByRequest.get(request);
+  }
+
+  @Override
+  public Call newCall(Request request) {
+    Context callingContext = Context.current();
+    contextsByRequest.put(request, callingContext);
```

Comment:
rather than trying to detect whether there is the right kind of `tag` method, it's definitely simpler to just use the cache here always.

Reply:
Nice catch I guess we need to build a copy of the Request even when not using the `tag` method.

Reply:
fixed